### PR TITLE
feat(hatchery/vsphere): refactor and tests

### PR DESCRIPTION
**Pull request by fsamin**
Signed-off-by: francois  samin <francois.samin@corp.ovh.com>

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15540.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15543.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Review by sguiheux — APPROVED** [review]

**Review by yesnault — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15551.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package and Analyze
  * Docker ✔ 
  * Sonar ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Review by richardlt — APPROVED** [review]

**Comment by fsamin** [comment]
don't merge now

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15596.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15604.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Review by sguiheux — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15608.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/hatchery/vsphere/hatchery.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+// ApplyConfiguration apply an object of type HatcheryConfiguration after checking it
+func (h *HatcheryVSphere) ApplyConfiguration(cfg interface{}) error {
+	if err := h.CheckConfiguration(cfg); err != nil {
+		return err
+	}
+
+	var ok bool
+	h.Config, ok = cfg.(HatcheryConfiguration)
+	if !ok {
+		return fmt.Errorf("Invalid configuration")
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:64]:
```suggestion
		return sdk.WithStack(fmt.Errorf("invalid configuration"))
```

File: engine/hatchery/vsphere/hatchery.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	h.Common.Common.ServiceName = h.Config.Name
+	h.Common.Common.ServiceType = sdk.TypeHatchery
+	h.HTTPURL = h.Config.URL
+	h.MaxHeartbeatFailures = h.Config.API.MaxHeartbeatFailures
+	var err error
+	h.Common.Common.PrivateKey, err = jwt.ParseRSAPrivateKeyFromPEM([]byte(h.Config.RSAPrivateKey))
+	if err != nil {
+		return fmt.Errorf("unable to parse RSA private Key: %v", err)
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:74]:
```suggestion
		return sdk.WithStack(fmt.Errorf("unable to parse RSA private Key: %v", err))
```

File: engine/hatchery/vsphere/hatchery.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+	m := h.NewMonitoringStatus()
+	m.AddLine(sdk.MonitoringStatusLine{Component: "Workers", Value: fmt.Sprintf("%d/%d", len(h.WorkersStarted(ctx)), h.Config.Provision.MaxWorker), Status: sdk.MonitoringStatusOK})
+	return m
+}
+
+// CheckConfiguration checks the validity of the configuration object
+func (h *HatcheryVSphere) CheckConfiguration(cfg interface{}) error {
+	hconfig, ok := cfg.(HatcheryConfiguration)
+	if !ok {
+		return fmt.Errorf("Invalid hatchery vsphere configuration")
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:91]:
```suggestion
		return sdk.WithStack(fmt.Errorf("invalid hatchery vsphere configuration"))
```

File: engine/hatchery/vsphere/hatchery.go [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	if err := hconfig.Check(); err != nil {
+		return fmt.Errorf("Invalid hatchery vsphere configuration: %v", err)
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:95]:
```suggestion
		return sdk.WrapError(err, "invalid hatchery vsphere configuration")
```

File: engine/hatchery/vsphere/hatchery.go [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	if hconfig.VSphereUser == "" {
+		return fmt.Errorf("vsphere-user is mandatory")
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:99]:
```suggestion
		return sdk.WithStack(fmt.Errorf("vsphere-user is mandatory"))
```

File: engine/hatchery/vsphere/hatchery.go [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	if hconfig.VSphereEndpoint == "" {
+		return fmt.Errorf("vsphere-endpoint is mandatory")
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:103]:
```suggestion
		return sdk.WithStack(fmt.Errorf("vsphere-endpoint is mandatory"))
```

File: engine/hatchery/vsphere/hatchery.go [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	if hconfig.VSpherePassword == "" {
+		return fmt.Errorf("vsphere-password is mandatory")
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:107]:
```suggestion
		return sdk.WithStack(fmt.Errorf("vsphere-password is mandatory"))
```

File: engine/hatchery/vsphere/hatchery.go [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	if hconfig.VSphereDatacenterString == "" {
+		return fmt.Errorf("vsphere-datacenter is mandatory")
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:111]:
```suggestion
		return sdk.WithStack(fmt.Errorf("vsphere-datacenter is mandatory"))
```

File: engine/hatchery/vsphere/hatchery.go [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+	}
+
+	if hconfig.IPRange != "" {
+		_, err := sdk.IPinRanges(context.Background(), hconfig.IPRange)
+		if err != nil {
+			return fmt.Errorf("flag or environment variable openstack-ip-range error: %v", err)
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:117]:
```suggestion
			return sdk.WrapError(err, "flag or environment variable openstack-ip-range")
```

File: engine/hatchery/vsphere/hatchery.go [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+		log.Error(ctx, "unable to get get vm template %q: %v", m.Name, err)
+		return true
+	}
+	if model.Config == nil || model.Config.Annotation == "" {
+		return true
+	}
+
+	var annot annotation
+	if err := json.Unmarshal([]byte(model.Config.Annotation), &annot); err != nil {
+		return true
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:206]: log error

File: engine/hatchery/vsphere/hatchery.go [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+		if strings.Contains(strings.ToLower(s.Name), strings.ToLower(model.Name)) {
+			x++
+		}
+	}
+	log.Debug(ctx, "WorkersStartedByModel> %s : %d", model.Name, x)
+
+	return x
+}
+
+// WorkersStarted returns the number of instances started but
```
Comment (richardlt) [inline engine/hatchery/vsphere/hatchery.go:240]:
```suggestion
// WorkersStarted returns the list of started workers
```

File: engine/hatchery/vsphere/ip.go [hunk 12]
```diff
@@ -2,71 +2,59 @@ package vsphere
...
+	h.reservedIPAddresses = append(h.reservedIPAddresses, ip)
+	h.IpAddressesMutex.Unlock()
 
-	ipToBook := freeIP[rand.Intn(len(freeIP))]
-	infos := ipInfos{
-		workerName:     workerName,
-		dateLastBooked: time.Now(),
-	}
-	ipsInfos.ips[ipToBook] = infos
+	go func() {
```
Comment (richardlt) [inline engine/hatchery/vsphere/ip.go:52]: Can we wait syncro, its strange to have this kind of code that can change the hatchery at any moment, no ?

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15609.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/hatchery/vsphere/vsphere.go [hunk 13]
```diff
@@ -2,311 +2,392 @@ package vsphere
...
-	cfg.Token = sConfig.API.Token
-	cfg.InsecureSkipVerifyTLS = sConfig.API.HTTP.Insecure
-	cfg.RequestSecondsTimeout = sConfig.API.RequestTimeout
-	return cfg, nil
+	return finder, nil
 }
 
-func (h *HatcheryVSphere) GetLogger() *logrus.Logger {
-	return h.ServiceLogger
+func (c *vSphereClient) ListVirtualMachines(ctx context.Context) ([]mo.VirtualMachine, error) {
```
Comment (richardlt) [inline engine/hatchery/vsphere/vsphere.go:68]: I don't see any filter on CDS created ressource, for openstack we check if a label hatchery_name exists.

File: sdk/hatchery/starter.go [hunk 14]
```diff
@@ -226,28 +243,30 @@ func spawnWorkerForJob(ctx context.Context, h Interface, j workerStarterRequest)
...
+	if remainingLength <= 5 {
+		return "", errors.New("modelName too long")
 	}
-	if len(hName) > 10 {
-		hName = ""
-	}
-	workerName = fmt.Sprintf("%s%s%s-%s", prefix, hName, modelName, random)
+
+	random := namesgenerator.GetRandomNameCDSWithMaxLength(remainingLength)
+	workerName := fmt.Sprintf("%s-%s-%s", nameFirstPart, random, hatcheryName)
```
Comment (richardlt) [inline sdk/hatchery/starter.go:262]: Look like random is added here and also line268

File: sdk/hatchery/starter_test.go [hunk 15]
```diff
@@ -18,28 +20,29 @@ func Test_generateWorkerName(t *testing.T) {
...
-			args: args{hatcheryName: "p999-prod", isRegister: true, model: "shared.infra-rust-official-1.41"},
-			want: "register-p999-prod-shared-infra-ru",
+			args: args{hatcheryName: "p999-prod", isRegister: true, model: "rust-official-1.41"},
+			want: "register-rust-official-1-41-",
 		},
 		{
 			name: "simple special char",
 			args: args{hatcheryName: "p999/prod", isRegister: true, model: "shared.infra-rust-official-1.41"},
-			want: "register-p999-prod-shared-infra-ru",
+			want: "register-shared-infra-rust-official-1-41-",
```
Comment (richardlt) [inline sdk/hatchery/starter_test.go:29]: ended by - is not really cool

**Review by richardlt — CHANGES_REQUESTED** [review]

File: engine/hatchery/vsphere/ip.go (hunk 12 above)
Comment (fsamin) [reply engine/hatchery/vsphere/ip.go:52]: same weird cache implementation in all the vsphere hatchery

File: engine/hatchery/vsphere/vsphere.go (hunk 13 above)
Comment (fsamin) [reply engine/hatchery/vsphere/vsphere.go:68]: this part of the code is a kind of low level vSphere client

Comment (fsamin) [reply engine/hatchery/vsphere/vsphere.go:68]: we are filtering CDS VM elsewhere with the function `getVirtualMachineCDSAnnotation`

File: sdk/hatchery/starter.go (hunk 14 above)
Comment (fsamin) [reply sdk/hatchery/starter.go:262]: yes

**Review by richardlt — APPROVED** [review]

**Review by yesnault — APPROVED** [review]

**Review by sguiheux — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15622.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15622.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15627.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Docs -   
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#15683.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Pull request merged**